### ISSUE TITLE
Low performance on trivial kernels for CUDA backend

### ISSUE THREAD

**Issue by Oblomov** [pocl/pocl#672]
The pocl CUDA backend achieves only about half the performance of the NVIDIA proprietary ICD, even when running trivial kernels.

Test cases illustrating the issue can be found [in this git repository](http://www.dmi.unict.it/~bilotta/gpgpu/svolti/aa201819/.git/). Clone it, go into the `opencl` directory, build with `make`, and run any of the examples, e.g.:

```
./vecsum $((16*1024*1024))
./vecsum2 $((16*1024*1024))
./vecsum_vec $((16*1024*1024)) 4
```

The (index of the) platform and device to use can be selected by setting the environment variables `OCL_PLATFORM` and `OCL_DEVICE`, otherwise the programs will use device 0 on platform 0.

On my GTX 750M, the NVIDIA proprietary driver will run the init kernel in ~2.3ms, and the sum kernel in 3.9, 3.2 and 3.1 ms respectively (cfr the examples above). The pocl CUDA backend will run init kernel in ~3.3ms, and the sum kernel in 5.3, 5.6 and 5.3 ms respectively.

pocl is built from the current `master` and I've tried both LLVM 5 and LLVM 7. The NVIDIA driver is the one shipping with Debian unstable, version 390.87-2. CUDA is also installed from distro packages, version 9.1.85-4+b1

**Comment by Oblomov** [comment]
Additional information: I tried running under `nvprof` to make sure that this wasn't just an issue with how timings were taken, and the results are pretty consistent between what gets reported by the OpenCL API and what gets reported by `nvprof`. For example, for the `vecsum_vec` case I get, from the OpenCL API:

```
init: 2.74189ms	48.9508GB/s
sum: 5.33694ms	37.7232GB/s
```

whereas `nvprof` tells me:

```
                   18.99%  5.3316ms         1  5.3316ms  5.3316ms  5.3316ms  vecsum4
                    9.67%  2.7144ms         1  2.7144ms  2.7144ms  2.7144ms  vecinit
```

(The timings for the memory transfers, read/map/unmap, are way off, but that's a different issue.)

So there is some issue with the actual kernel code and/or kernel launch. Something for @jrprice to look at I guess?

**Comment by jrprice** [comment]
I notice you're passing `NULL` for the work-group size in your `clEnqueueNDRangeKernel` calls, which might lead to pocl using a different work-group size than the NVIDIA driver (which could have a big impact). Could you try explicitly picking a work-group size for both drivers to see if that makes a difference?

**Comment by Oblomov** [comment]
I have tested with fixed work-group sizes and the performance difference is still there. For example, setting the lws to 256, the timing for vecsum are 3.9ms (NVIDIA) vs 5.3ms (pocl CUDA). I've pushed an `lws` branch to the git repository where you can choose the LWS by environment variable `LWS_SUM`.

Both branches have been updated also with a `vecsmooth` and `vecsmooth_lmem` programs. Interesting, for `vecsmooth` the automatic selection _does_ make a difference (NVIDIA seems to pick something like 256 that behaves better than the 1024 (I think) picked by pocl CUDA), since forcing the LWS in this case gives results which are _much_ closer in performance. (The results returned by

**Comment by Oblomov** [comment]
(Sorry, closed by mistake due to a misclick.)

**Comment by inducer** [comment]
By comparing the PTX emitted by either compiler, @nchristensen pinpointed a potential root of this (or a related) problem. The Nvidia compiler appears to emit pointer arithmetic in local memory as 32-bit, whereas pocl (and LLVM) appear to emit it as 64-bit. @nchristensen also found [this option](https://github.com/llvm-mirror/clang/commit/4f482660272cb79d5bec53291b043b0b9b19c29e) which was added to LLVM, seemingly by someone at Google for their LLVM->PTX compiler, and seemingly to address this same performance issue. So resolving (at least part of) this might be as simple as passing `--nvptx-short-ptr` during compilation. We have been (and are still) trying to come up with a way of passing the flag to the compilation process to test the theory, but we're currently held up in that by #752. We encountered this when investigating the performance of a more complex kernel, but it seems that the common theme here might be the use of shared/local memory.

**Comment by Oblomov** [comment]
My code doesn't use local memory though, it's only very straightforward writes or read, sum, write, embarrassingly parallel. Maybe there's some additional simplifications that can be done? One possibility could be the use of 32-bit vs 64-bit integers for the work-group and work-item indices and sizes.

**Comment by pjaaskel** [comment]
@inducer you can pass device-specific flags to the build via the init_build driver hook. See: https://github.com/pocl/pocl/blob/master/lib/CL/devices/tce/tce_common.cc#L732 for an example. This passes it to the clang though, not to the code gen phase.

**Comment by isuruf** [comment]
Is pocl not using standard LLVM passes that clang is using? I see pocl producing kernels with code similar to what `clang -O1 -x cl` produces and much different to `clang -O2 -x cl`.

**Comment by pjaaskel** [comment]
Yes it is. In frontend it uses -O1, IIRC (pocl_llvm_build.cc) to produce the program.bc. Then in pocl_llvm_wg.cc it uses also the standard LLVM optimizations (-O3) to optimize the final bitcode (parallel.bc). I'm not sure if the CUDA backend bypasses the workgroup function generation completely.

**Comment by pjaaskel** [comment]
Is the CUDA BE nowadays faster?

**Comment by inducer** [comment]
I'd call it pretty usable. We're using it as a "daily driver" in a reasonably large research center. Maybe @matthiasdiener has some recent data sitting around?
